### ISSUE TITLE
Incremental floating point is much slower than one-shot on certain short problems

### ISSUE THREAD

**Issue by arotenberg** [Z3Prover/z3#1459]
Z3 version 4.6.0 - 64 bit on Windows 7

This script takes about 14 seconds on my machine. However, if you comment out the `(push 1)` line, it's basically instant.

    (set-logic QF_FP)
    (declare-const a Float64)
    (declare-const b Float64)
    (assert (= b (fp.sqrt RNE a)))
    (push 1)
    (assert (= a (fp #b0 #b10000000000 #x0000000000000)))
    (check-sat)

This is weird because it seems like, even in incremental mode, Z3 should be able to constant fold this problem, or at least unit propagate it out of existence.

Similarly, this script takes about 2 seconds with the `(push 1)` line commented out, but with it included, I ran it for over 6 minutes before giving up. Z3's one-shot solution is to set `a` to a negative number and avoid most of the problem; I'm not sure why this wouldn't also be possible in the incremental version.

    (set-logic QF_FP)
    (declare-const a Float64)
    (declare-const b Float64)
    (assert (= b (fp.sqrt RNE a)))
    (push 1)
    (check-sat)

**Comment by wintersteiger** [comment]
The "one-shot" solver translates everything into bit-vectors, bit-blasts, then runs a specialized SAT-only solver. The incremental version uses a completely different (SMT) solver which supports theory combination, supports incrementality, and which is lazy in terms of float-to-bv conversion and bit-blasting, and uses different heuristics. I.e., it's fully expected that it behaves differently. I don't have time to look at the code right now, but I think the "obvious" propagation is not performed here (for any theory), because it didn't behave well enough on other problems. Should be fixable in other ways though.

**Comment by arotenberg** [comment]
The non-incremental solver is currently so much faster on this type of problem than the incremental solver that our team is actually considering whether our client program should have an option to write slight variants of the full problem over and over again to the non-incremental solver instead of using push/pop (or have options to run the incremental and non-incremental approaches in parallel, or dovetailing between the incremental and non-incremental approaches).

Is there a way to ask Z3 to implement push/pop using the non-incremental bit blasting solver by saving the history of assertions in memory and replaying them, rather than having our client program do this manually?

**Comment by wintersteiger** [comment]
If you want to have full control over it, you should implement that in the client. Z3 also offers similar ways to do that though: use tactics instead of check-sat, or construct a solver object via the API (push/pop is implemented similarly for them). You can also try adding assumption literals which allow you en/disable assertions later on if there aren't too many scopes. 

In general, it's not clear that push/pop is the right solution for you; if your queries are as small as the example posted here, it's probably not going to pay off. In the past, some users also adopted a strategy where they would push/pop for some time, but after some number of query they would throw it all away and re-build all the assertions, which worked better for them than always vs never pushing.

**Comment by wintersteiger** [comment]
Until we have a more general fix available, you can try using this custom tactic instead of `(check-sat)`:

```(check-sat-using (then simplify propagate-values fpa2bv simplify bit-blast simplify propagate-values sat))```


That's essentially what the one-shot solver runs and it will treat push/pop by copying all assertions that are in scope into a separate solver each time.

**Comment by arotenberg** [comment]
Thank you, we will try that.

Unfortunately, our queries are in general much larger than the examples posted here. I just reduced the examples to the simplest cases I could find for the bug report.

When you say "assumption literals", are you referring to `(check-sat-assuming)`?

**Comment by wintersteiger** [comment]
Yes, that's what I was referring to!

**Comment by rainoftime** [comment]
Here is an alternative approach.
- You use a solver `sol_store` to store the update the assertions with `add`, `push/pop`.
- But when checking the assertions, you can create a new temporal solver, `sol_tmp`. Then add the assertions in `sol_store` to the solver `sol_tmp`.
     
    ~~~~
      context c;
      sol_sotre(c);
      sol_store.add(e1);
      sol_store.add(e2);
      sol_tmp(c);
      sol_tmp.add(mk_and(sol_sotre.assertions());
     ~~~~

**Comment by wintersteiger** [comment]
@rainoftime The overhead really depends on the number and sizes of the expressions, but your suggestion is essentially what Z3 does when you create a solver from a tactic and then push/pop on it (or `check-sat-using` in SMT2 files). Using assertion literals you can add all constraints up front, but then disable some of them later by setting the assumption literals to false. Which approach works better for you really depends on your problems though.

**Comment by wintersteiger** [comment]
Closing this as resolved. If this or similar issues come up again, please re-open or create a new issue.
